### ISSUE TITLE
Values treated as globals when file has no imports or exports

### ISSUE THREAD

**Issue by mjomble** [Microsoft/TypeScript#21109]
**TypeScript Version:**  2.6.2 and 2.7.0-dev.20180109

**Code**

All three files in the same dir.

a.ts
```ts
console.log(foo)
```

b.ts
```
const foo = 0
```

tsconfig.json
```json
{ "include": ["./*"] }
```

**Expected behavior:**
Since `b.ts` does not export `foo` and `a.ts` does not import it, running `tsc` in this dir should give an error like this:

```
a.ts(1,13): error TS2304: Cannot find name 'foo'.
```

For example, running `node a.js` results in `ReferenceError: foo is not defined`

**Actual behavior:**
No error is reported.

Each of the following changes results in the error being correctly reported:

* Removing `b.ts`
* Removing (or renaming) the declaration of `foo`
* Adding any imports or exports to `b.ts` (for example, exporting `foo` itself)
* Wrapping the `const foo = 0` line in a function so it's no longer a top-level declaration

However, even adding other imports or exports to `a.ts` doesn't trigger the error, as long as `b.ts` does not contain any imports or exports.

So my guess is that the current behavior is: if a file contains no imports or exports, all of its top-level declarations are considered globally available to other files.

Is this a bug or intentional?

**Comment by mhegazy** [comment]
> Is this a bug or intentional?

This is behaving as intended. A file is considered as a module iff it has at least one top-level import or export.  consider adding `export {}` to disambiguate the meaning of the file.

**Comment by mjomble** [comment]
Interesting. I can see how this might be desirable when neither of the files (or no file in your project) is a module.

However, is it also intentional that declarations from non-modules are globally available in modules? Perhaps they should only be available in other non-modules.

It's easy to accidentally create a non-module when refactoring a larger file into multiple smaller ones and forgetting to add an `export`.

**Comment by mhegazy** [comment]
Modules introduce their own scope. Script files contribute to the global scope and global scope is visible everywhere.

**Comment by kitsonk** [comment]
> It's easy to accidentally create a non-module when refactoring a larger file into multiple smaller ones and forgetting to add an `export`.

Are modules that neither import anything or export anything that common?  If the only thing a module does is augment the global scope, then it isn't really a module...  A module should have some sort of public API surface (or be a _final_ module that imports other functionality to perform an action).

**Comment by mjomble** [comment]
> Are modules that neither import anything or export anything that common?

Not that I know of, which is why I only create them by accident.
My intention is to split a large module into smaller modules, each with their own exports.
In the process, I need to add the `export` keyword to declarations that used to be internal.
If by accident I forget to add the keyword to all declarations in a file (for example a small file that only contains 1-2 declarations) it can accidentally end up being a non-module.
Even though it was not my intention to create a module without exports.

I guess if mixing modules and non-modules is supported by design in TypeScript, I should use other tools to ensure that all the files in my project are modules.

**Comment by mjomble** [comment]
There's now a clarification for this in the documentation (at least in the repo), so I'm closing the issue.
